include/seqan3/alignment/pairwise/align_result.hpp

```diff
@@ -34,210 +34,154 @@
 
 /*!\file
  * \brief Provides seqan3::align_result.
- * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ * \author Jörg Winkler <j.winkler AT fu-berlin.de>
+ *
+ * \details
+ * Objects of this class are the result of an alignment computation.
+ * It always contains and alignment identifier and the resulting score.
+ * Optionally – if the user requests – also the begin and end positions within
+ * the sequences and the trace can be calculated. When accessing a field that
+ * has not been calculated an std::bad_optional_access exception is thrown.
  */
 
 #pragma once
 
-#include <functional>
 #include <optional>
-#include <tuple>
-
-#include <seqan3/core/metafunction/template_inspection.hpp>
 
 namespace seqan3
 {
 
-/*!\brief Keys for different alignment results.
+/*!\brief Stores the alignment results and gives access to score, traceback and the begin and end coordinates.
  * \ingroup pairwise
+ * \tparam id_t         The type for the alignment identifier.
+ * \tparam score_t      The type for the resulting score.
+ * \tparam coordinate_t The type for both the begin and end coordinates, can be omitted.
+ * \tparam trace_t      The type for the alignment trace, can be omitted.
  */
-enum struct align_result_key : uint8_t
+template <typename id_t, typename score_t, typename coordinate_t = bool, typename trace_t = bool>
+class align_result
 {
-    id,     // cannot be set by the config
-    score,  // report the score
-    end,    // report the end position and score
-    begin,  // report the begin and end position and score
-    trace   // report the full trace and begin and end and score
-};
-
-/*!\brief Stores the alignment results and offers a tuple-like interface.
- * \extends std::tuple
- * \ingroup pairwise
- * \tparam output_type_list_t seqan3::type_list over the contained results.
- */
-template <typename output_type_list_t>
-struct align_result //!\cond
-    : public detail::transfer_template_args_onto_t<output_type_list_t, std::tuple>
-    //!\endcond
-{
-
-    static_assert(meta::size<output_type_list_t>::value >= 2,
-                  "Logic error: This class requires at least a type list with two template arguments "
-                  "for the id and the alignment score");
-
-    //!\brief The base tuple type.
-    using base_type = detail::transfer_template_args_onto_t<output_type_list_t, std::tuple>;
-
-    //!\brief Inheriting the constructors of the base type.
-    using base_type::base_type;
+private:
+    //! \brief The alignment identifier.
+    id_t id;
+    //! \brief The alignment score.
+    score_t score;
+    //! \brief The end coordinate of the alignment.
+    std::optional<coordinate_t> end_coordinate;
+    //! \brief The begin coordinate of the alignment.
+    std::optional<coordinate_t> begin_coordinate;
+    //! \brief The alignment trace, i.e. the actual base pair matching.
+    std::optional<trace_t> trace;
+
+public:
+    /*!\brief Constructor with all available fields.
```

Comment:
From the design I suggest a different approach.
Provide a `detail::alignment_result_value_type` which is an aggregate type (no constructors).
It has deduction guides for only the first, then the first and the second etc.

This is what I mean, but with more descriptive names.
The non set types will default to `void *`
```cpp
template <typename t1, typename t2 = void *>
struct alignment_result_value_type
{
    t1 v1{};
    t2 v2{};
};

template <typename t1>
alignment_result_value_type(t1) -> alignment_result_value_type<t1>;

template <typename t1, typename t2>
alignment_result_value_type(t1, t2) -> alignment_result_value_type<t1, t2>;

int main()
{
    alignment_result_value_type o{4};

    alignment_result_value_type o2{4, .5};
}
```

The alignment result class is templatized only over this struct, which we can constrain then with `is_specialization_of` as well. And there is only one constructor. The access functions could then static assert if the requested type is a `void *` instead of throwing. I realized yesterday while working on the algorithm implementation, that the type of the result must be and can determined anyway. 
The default constructor could deduce to a value type that only has void *.

Reply:
Do you want to have the result members public or as private members with getters?

Reply:
Hm isn't this also too complicated? What about this:
```cpp
#include <optional>

template<typename t1, typename t2 = void *>
struct S
{

    S(t1 i2 = t1{}, t2 f2 = t2{}): i{i2}, f{f2}
    {} 

    t1 get_i()
    {
        static_assert(!std::is_same_v<t1, void *>);
        return i;
    }

    t2 get_f()
    {
        static_assert(!std::is_same_v<t2, void *>);
        return f;
    }

private:
    t1 i{};
    t2 f{};
};

int main()
{
    S mys{1, 1.0}; // for S mys{1}; the static assertion fails

    mys.get_f();
}

```
No multiple constructors, no multiple type deductions. no extra detail struct, static asserts for getter functions.

Reply:
All fine but, in your example all types are exposed in the user interface, rather then just one type.
Second, it forbids me to use `designated initializers` in the future. So within the alignment computation I would love to just do: 
```cpp
alignment_result{result_value_type{.id = 0, .score = sc, .end_coordinage = {30, 12}}}
``` 
Third, I can pass in the arguments in a wrong order, which I could prevent with strong types. In the above solution I can still mix the order, but at least I have a concrete naming for the variables and don't need to introduce all the strong types. 
Fourth, the implementation of the values is decoupled from the access object, which means I can reassign the values in the alignment computation. The user should not be able to do this.

Reply:
I see, I might still ask you out of interest later on as I don't understand all your points.
I just wanted to make a suggestion since the issue seemed easy enough at a first glance but as it looks like there is more to it I'll leave this design decision to you.

Reply:
I implemented the proposed design, so you can have a look and compare.